File: Examen1/2026_1/PlantillaC#/TestSoft/TestSoftPersistance/Preguntas/Impl/PreguntaImpl.cs
using SoftProgDBManager;
using TestSoftModel.Pregunta;
using TestSoftPersistance.Preguntas.DAO;
using System.Data;
using System.Data.Common;

namespace TestSoftPersistance.Preguntas.Impl
{
    public class PreguntaImpl : PreguntaDAO
    {
        // SP: insertar_pregunta(IN p_enunciado, OUT p_id)
        public int Insertar(Pregunta pregunta)
        {
            DbParameter[] parametros = new DbParameter[2];
            parametros[0] = DBManager.Instance.CreateParam("p_enunciado", DbType.String, pregunta.Enunciado, ParameterDirection.Input);
            parametros[1] = DBManager.Instance.CreateParam("p_id", DbType.Int32, null, ParameterDirection.Output);

            DBManager.Instance.EjecutarProcedimiento("insertar_pregunta", parametros);
            pregunta.Id = Convert.ToInt32(parametros[1].Value);
            return pregunta.Id;
        }

        // SP: modificar_pregunta(IN p_id, IN p_enunciado)
        public int Modificar(Pregunta pregunta)
        {
            DbParameter[] parametros = new DbParameter[2];
            parametros[0] = DBManager.Instance.CreateParam("p_id", DbType.Int32, pregunta.Id, ParameterDirection.Input);
            parametros[1] = DBManager.Instance.CreateParam("p_enunciado", DbType.String, pregunta.Enunciado, ParameterDirection.Input);
            return DBManager.Instance.EjecutarProcedimiento("modificar_pregunta", parametros);
        }

        // SP: eliminar_pregunta(IN p_id)
        public int Eliminar(int id)
        {
            DbParameter[] parametros = new DbParameter[1];
            parametros[0] = DBManager.Instance.CreateParam("p_id", DbType.Int32, id, ParameterDirection.Input);
            return DBManager.Instance.EjecutarProcedimiento("eliminar_pregunta", parametros);
        }

        public Pregunta BuscarPorId(int id)
        {
            return null; // No hay SP buscar_pregunta_por_id
        }

        // SP: listar_preguntas()
        public List<Pregunta> ListarTodos()
        {
            List<Pregunta> lista = new List<Pregunta>();
            using DbDataReader lector = DBManager.Instance.EjecutarProcedimientoLectura("listar_preguntas", null);
            while (lector.Read())
            {
                Pregunta p = new Pregunta();
                p.Id = lector.GetInt32(lector.GetOrdinal("id"));
                p.Enunciado = lector.GetString(lector.GetOrdinal("enunciado"));
                lista.Add(p);
            }
            return lista;
        }
    }
}

File: Examen1/2026_1/Plantillas/TestSoft/TestSoftPersistance/src/main/java/pe/edu/pucp/testsoft/persistance/pregunta/impl/PreguntaImpl.java
package pe.edu.pucp.testsoft.persistance.pregunta.impl;

import pe.edu.pucp.testsoft.config.DBManager;
import pe.edu.pucp.testsoft.model.pregunta.Pregunta;
import pe.edu.pucp.testsoft.persistance.pregunta.dao.PreguntaDAO;

import java.sql.Types;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class PreguntaImpl implements PreguntaDAO {
    private final DBManager dbManager = DBManager.getInstance();

    // SP: insertar_pregunta(IN p_enunciado, OUT p_id)
    @Override
    public int insertar(Pregunta pregunta) {
        Map<Integer, Object> entrada = new HashMap<>();
        Map<Integer, Object> salida = new HashMap<>();
        entrada.put(1, pregunta.getEnunciado());
        salida.put(2, Types.INTEGER);

        dbManager.ejecutarProcedimiento("insertar_pregunta", entrada, salida);
        pregunta.setId((int) salida.get(2));
        return pregunta.getId();
    }

    // SP: buscar_alumno_por_id... no, para pregunta no hay SP de buscar por id
    // pero lo implementamos por si acaso con listar y filtrar, o se puede crear el SP
    @Override
    public Pregunta buscarPorId(int id) {
        // Nota: se podría crear un SP buscar_pregunta_por_id si se necesita
        return null;
    }

    // SP: listar_preguntas()
    @Override
    public List<Pregunta> listarTodos() {
        List<Pregunta> lista = null;
        try (DBManager.ResultadoConsulta rc = dbManager.ejecutarProcedimientoLectura(
                "listar_preguntas", null)) {
            while (rc.getRs().next()) {
                if (lista == null) lista = new ArrayList<>();
                Pregunta p = new Pregunta();
                p.setId(rc.getRs().getInt("id"));
                p.setEnunciado(rc.getRs().getString("enunciado"));
                lista.add(p);
            }
        } catch (Exception ex) {
            System.out.println("Error al listar preguntas: " + ex.getMessage());
        }
        return lista;
    }

    // SP: modificar_pregunta(IN p_id, IN p_enunciado)
    @Override
    public int modificar(Pregunta pregunta) {
        Map<Integer, Object> entrada = new HashMap<>();
        entrada.put(1, pregunta.getId());
        entrada.put(2, pregunta.getEnunciado());
        return dbManager.ejecutarProcedimiento("modificar_pregunta", entrada, null);
    }

    // SP: eliminar_pregunta(IN p_id)
    @Override
    public int eliminar(int id) {
        Map<Integer, Object> entrada = new HashMap<>();
        entrada.put(1, id);
        return dbManager.ejecutarProcedimiento("eliminar_pregunta", entrada, null);
    }

}
